### PR TITLE
[OSDEV-1335] Moderation. Deploy new moderation events Logstash pipeline

### PR DESCRIPTION
**Pull request by vladsha-dev**
- The new "moderation events" Logstash pipeline has been configured and implemented to collect moderation event data from the current PostgreSQL database and save it to OpenSearch. This setup allows for fast searches on the moderation events data.
- Fixed the assertion in the test for the `country.rb` filter of the "production locations" Logstash pipeline. The main issue was with the evaluation of statements in the Ruby block. Since only the last statement is evaluated in a Ruby block, all the checks were grouped into one chain of logical statements and returned as a `result` variable at the end.
- Explicitly set the number of shards and the number of replicas for the "production locations" and "moderation events" OpenSearch indexes. Based on the OpenSearch documentation, a storage size of 10–30 GB is preferred for workloads that prioritize low search latency. Additionally, having too many small shards can unnecessarily exhaust memory by storing excessive metadata. Currently, the "production locations" index utilizes 651.9 MB in the Production env, including replicas, while the "moderation events" index is empty. This indicates that one shard and one replica should be sufficient for the "production locations" and "moderation events" indexes.

### CODE REVIEW

**Review by VadimKovalenkoSNF — CHANGES_REQUESTED** [review]
Ask for change about OpenSearch indexation.

File: deployment/clear_opensearch/clear_opensearch.sh.tpl [hunk 1]
```diff
@@ -1,7 +1,10 @@
 #!/bin/bash
 echo -e "\nDelete OpenSearch indexes\n"
 curl -X DELETE https://$OPENSEARCH_DOMAIN/production-locations --aws-sigv4 "aws:amz:eu-west-1:es" --user "$AWS_ACCESS_KEY_ID:$AWS_SECRET_ACCESS_KEY"
-echo -e "\nRemove lock file\n"
+curl -X DELETE https://$OPENSEARCH_DOMAIN/moderation-events --aws-sigv4 "aws:amz:eu-west-1:es" --user "$AWS_ACCESS_KEY_ID:$AWS_SECRET_ACCESS_KEY"
```
[the remark this answers is not in the document]
Comment (roman-stolar) [reply deployment/clear_opensearch/clear_opensearch.sh.tpl:4]: Useful

File: deployment/clear_opensearch/clear_opensearch.sh.tpl [hunk 2]
```diff
@@ -1,7 +1,10 @@
...
+
+echo -e "\nRemove the JDBC input lock files from the EFS connected to Logstash\n"
 sudo mount -t efs -o tls,accesspoint=$EFS_AP_ID $EFS_ID:/ /mnt
-sudo rm /mnt/logstash_jdbc_last_run
+sudo rm /mnt/production_locations_jdbc_last_run
+sudo rm /mnt/moderation_events_jdbc_last_run
```
[the remark this answers is not in the document]
Comment (roman-stolar) [reply deployment/clear_opensearch/clear_opensearch.sh.tpl:9]: Useful, but optional

File: src/logstash/indexes/moderation_events.json [hunk 3]
```diff
@@ -0,0 +1,64 @@
+{
+  "index_patterns": [
+    "moderation-events*"
+  ],
```
[the remark this answers is not in the document]
Comment (roman-stolar) [reply src/logstash/indexes/moderation_events.json:4]: I'll resolved.

File: src/logstash/scripts/moderation_events/country.rb [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        # Country code Alpha-2 data.
+        alpha_2_country_code = cleaned_data_value['country_code']
+
+        # Define the path to the JSON file.
+        json_countries_file_path = File.expand_path('../../../static_data/countries.json', __FILE__)
+
+        # Read and parse the JSON file.
+        json_countries_data = File.read(json_countries_file_path)
+        countries = JSON.parse(json_countries_data)
+        exact_country = countries[alpha_2_country_code]
```
[the remark this answers is not in the document]
Comment (roman-stolar) [reply src/logstash/scripts/moderation_events/country.rb:18]: Useful, but optional

**Review by roman-stolar — APPROVED** [review]
In general LGTM, just need to covered @VadimKovalenkoSNF and some @coderabbit comments.

File: src/logstash/indexes/moderation_events.json [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        "contributor_id": {
+          "type": "keyword"
+        },
+        "contributor_name": {
+          "type": "keyword"
+        },
+        "claim_id": {
+          "type": "keyword"
+        },
+        "request_type": {
```
[the remark this answers is not in the document]
Comment (vladsha-dev) [reply src/logstash/indexes/moderation_events.json:23]: @VadimKovalenkoSNF 
It is already for filters. The keyword is for exact search. Am I missing something?

[the remark this answers is not in the document]
Comment (vladsha-dev) [reply src/logstash/indexes/moderation_events.json:23]: Did you mean `terms` search? I checked, and it works for that type of OpenSearch query.
I used the following query:
```
{
    "query": {
        "bool": {
            "must": [
                {
                    "terms": {
                        "source": [
                            "API",
                            "SLC"
                        ]
                    }
                }
            ]
        }
    }
}
```

[the remark this answers is not in the document]
Comment (VadimKovalenkoSNF) [reply src/logstash/indexes/moderation_events.json:23]: Current query builder is not adapted to build filter queries like that. It should be:

```
{
    "query": {
        "bool": {
            "must": [
                {
                    "terms": {
                        "source.keyword": [
                            "API",
                            "SLC"
                        ]
                    }
                }
            ]
        }
    }
}
```

See (`source.keyword` vs `source`). Basically, the problem here is that we use query builder on top for `production-locations` mapping which introduces some limitations.

[the remark this answers is not in the document]
Comment (vladsha-dev) [reply src/logstash/indexes/moderation_events.json:23]: I believe this is an issue that should be addressed on the query builder side to make that code more adaptive, as it's not optimal to force OpenSearch to index data in both ways. This approach will lead to additional storage consumption within the OpenSearch instance.

[the remark this answers is not in the document]
Comment (VadimKovalenkoSNF) [reply src/logstash/indexes/moderation_events.json:23]: Yes, I think it is better idea to extend query builder like:
`OpenSearchQueryBuilder(OpenSearchQueryBuilderInterface) → ProductionLocationsQueryBuilder | ModerationEventsQueryBuilder`

[the remark this answers is not in the document]
Comment (vladsha-dev) [reply src/logstash/indexes/moderation_events.json:23]: Yeah

**Review by VadimKovalenkoSNF — DISMISSED** [review]
Approved, but as Roman mentioned, resolve coderabbit comments.

File: src/logstash/sql/sync_moderation_events.sql [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+	ame.uuid AS moderation_id,
+	ame.os_id AS os_id_value,
+	ame.contributor_id,
+	(
+	SELECT
+		name
+	FROM
+		api_contributor ac
+	WHERE
+		id = ame.contributor_id) AS contributor_name,
```
[the remark this answers is not in the document]
Comment (vladsha-dev) [reply src/logstash/sql/sync_moderation_events.sql:11]: Done

File: src/logstash/indexes/moderation_events.json [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+                  "type": "keyword"
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (vladsha-dev) [reply src/logstash/indexes/moderation_events.json:64]: Done

**Review by mazursasha1990 — CHANGES_REQUESTED** [review]

File: src/logstash/config/pipelines.yml [hunk 8]
```diff
@@ -0,0 +1,8 @@
+# This file is where pipelines are defined.
+# For more information on multiple pipelines, see the documentation:
+#   https://www.elastic.co/guide/en/logstash/8.15/multiple-pipelines.html
+
+- pipeline.id: production_locations
+  path.config: "/usr/share/logstash/pipeline/sync_production_locations.conf"
+- pipeline.id: moderation_events
+  path.config: "/usr/share/logstash/pipeline/sync_moderation_events.conf"
```
[the remark this answers is not in the document]
Comment (vladsha-dev) [reply src/logstash/config/pipelines.yml:8]: Done

File: src/logstash/config/logstash.yml [hunk 9]
```diff
@@ -0,0 +1 @@
+api.enabled: false
```
[the remark this answers is not in the document]
Comment (vladsha-dev) [reply src/logstash/config/logstash.yml:1]: Done

File: src/logstash/scripts/moderation_events/country.rb
[the remark this answers is not in the document]
Comment (vladsha-dev) [reply src/logstash/scripts/moderation_events/country.rb:1]: I deleted obvious comments but left several for clarity.

**Review by roman-stolar — APPROVED** [review]
LGTM

File: src/logstash/sql/sync_moderation_events.sql [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+FROM
+	api_moderationevent ame
+LEFT JOIN api_contributor AS ac
+ON
+	ac.id = ame.contributor_id
+WHERE
+	ame.updated_at > :sql_last_value
+	AND ame.updated_at < CURRENT_TIMESTAMP
+ORDER BY
+	ame.updated_at ASC
```
Comment (mazursasha1990) [inline src/logstash/sql/sync_moderation_events.sql:23]: Please add an empty line.

**Review by mazursasha1990 — CHANGES_REQUESTED** [review]

File: src/logstash/sql/sync_moderation_events.sql (hunk 10 above)
Comment (vladsha-dev) [reply src/logstash/sql/sync_moderation_events.sql]: Done

**Review by mazursasha1990 — APPROVED** [review]
LGTM

**Review by VadimKovalenkoSNF — APPROVED** [review]
Approved.

**Pull request merged**